File: erlang/seminars/tracer/src/tracer.erl
-module(tracer).

-export([tracer/2]).

-define(Black, {0,0,0}).
-define(White, {1,1,1}).

tracer(Camera, Objects) ->
    {W, H} = camera:size(Camera),
    Xs = lists:seq(1, W),
    Ys = lists:seq(1, H),
    [[trace(X, Y, Camera, Objects) || X <- Xs] || Y <- Ys].

trace(X, Y, Camera, Objects) ->
    Ray = camera:ray(X, Y, Camera),
    trace(Ray, Objects).

trace(Ray, Objects) ->
    case intersect(Ray, Objects) of
        {inf, _} ->
	    ?Black;
	{_, _} -> 
	    ?White
    end.

intersect(Ray, Objs) ->
   lists:foldl(fun(Obj, Sofar) -> 
		       {Dist, _} = Sofar,
		       case objects:intersect(Obj, Ray) of
			   {ok, D} when D < Dist ->
			       {D, Obj};
			   _ ->
			       Sofar
		       end
	       end,
	       {inf, no},
	       Objs).

File: exercises/tracer/src/tracer.ex
defmodule Tracer do

  @black {0, 0, 0}
  @white {1, 1, 1}

  def tracer(camera=%Camera{}, objects) do
    {w, h} = camera.size
    xs = Enum.to_list(1..w)
    ys = Enum.to_list(1..h)
    for y <- ys, do: for(x <- xs, do: trace(x, y, camera, objects))
  end

  def trace(x, y, camera, objects) do
    ray = Camera.ray(camera, x, y)
    trace(ray, objects)
  end

  def trace(ray, objects) do
    case intersect(ray, objects) do
      {:inf, _} ->
        @black

      {_, _} ->
        @white
    end
  end

  def intersect(ray, objects) do
    List.foldl(objects, {:inf, nil}, fn (object, sofar) ->
      {dist, _} = sofar

      case Object.intersect(object, ray) do
        {:ok, d} when d < dist ->
          {d, object}
        _ ->
          sofar
      end
    end)
  end

end
